[CMSP-995] one more notice

jazzsequence opened the pull request:
This PR adds another notice that appears globally (not just on the Pantheon Cache page) if your max-age value is less than the default. This notice goes away after you have seen it once. The type of the notice (error (red) or warning (yellow)) is determined by the `$max_age_rank` value, but the language of the notice does not change.

<img width="916" alt="Screenshot 2024-05-17 at 3 39 38 PM" src="https://github.com/pantheon-systems/pantheon-advanced-page-cache/assets/991511/bf19c3b0-255f-4f98-a10c-1bc83af799cf">

jazzsequence committed:
add a global admin notice that appears once

jazzsequence committed:
remove seconds from the notice

jazzsequence committed:
test the new notice

jazzsequence committed:
set dismissable depending on which notice we're showing

jazzsequence committed:
show the max age rank html comment

jazzsequence committed:
fix the conditionals for cs

jazzsequence committed:
update behat tests for new notice language

jazzsequence committed:
bugfix: don't update user meta that max_age was updated if it hasn't been

pwtyler commented on inc/admin-interface.php:
```diff
@@ -111,39 +111,61 @@ function admin_notice_old_mu_plugin() {
 		$max_age_rank > 0 &&
 		$current_max_age < WEEK_IN_SECONDS &&
-		$current_max_age !== 600
-	) {
-		// If the current max-age value has a rank of 3 or more (10 is the highest), we'll note that it's very low.
+		$current_max_age !== 600 &&
+		isset( $current_screen->id ) &&
+		'settings_page_pantheon-cache' === $current_screen->id
+		) {
+			// If the current max-age value has a rank of 3 or more (10 is the highest), we'll note that it's very low.
 		$very_low = $max_age_rank > 3 ? __( 'This is a very low value and may not be optimal for your site.', 'pantheon-advanced-page-cache' ) : '';
 		$message = sprintf(
 			// translators: %1$s is the current max-age, %2$d is the current max-age in seconds, %3$s is a message that displays if the value is very low, %44d is the recommended max age in seconds, %5$s is the humanized recommended max age, %6$s is debug information that is written to the HTML DOM but not displayed.
-			__( 'The cache max-age is currently set to %1$s (%2$s seconds). %3$s Consider increasing the cache max-age to at least %4$d seconds (%5$s).%6$s', 'pantheon-advanced-page-cache' ),
+			__( 'The cache max-age is currently set to %1$s. %2$s Consider increasing the cache max-age to at least %3$s.%4$s', 'pantheon-advanced-page-cache' ),
 			humanized_max_age(),
-			$current_max_age,
 			$very_low,
-			WEEK_IN_SECONDS,
 			humanized_max_age( true ),
 			sprintf( '<!-- Max Age Rank: %d -->', $max_age_rank )
 		);
+	}
+
+	// Global notice on all pages _except_ the Pantheon cache settings page.
+	if (
+		$max_age_rank > 0 &&
+		$current_max_age < WEEK_IN_SECONDS &&
+		$current_max_age !== 600 &&
+		( ! isset( $current_screen->id ) || 'settings_page_pantheon-cache' !== $current_screen->id )
+	) {
+		$message = sprintf(
+			// translators: %s is a link to the Pantheon GCDN configuration page.
+			__( 'Your site\'s cache max-age is set below the recommendation (1 week). Visit the <a href="%1$s">Pantheon GCDN configuration page</a> to update the setting.%2$s' ),
+			admin_url( 'options-general.php?page=pantheon-cache' ),
+			sprintf( '<!-- Max Age Rank: %d -->', $max_age_rank )
+		);
+		$dismissable = false;
+		update_user_meta( get_current_user_id(), 'pantheon_max_age_global_warning_notice', true );
+	}
```
at a glance I think these two if statements could be refactored a bit since you're checking ```$max_age_rank > 0 &&
		$current_max_age < WEEK_IN_SECONDS &&
		$current_max_age !== 600 ``` in both places. but i didn't attempt an approach

pwtyler commented on tests/phpunit/test-admin-interface.php:
```diff
@@ -271,4 +271,20 @@ function test_max_age_updated_admin_notice() {
 		$this->assertEquals( 1, get_user_meta( $current_user_id, 'pantheon_max_age_updated_notice', true ) );
 	}
+
+	/**
+	 * Test that the user meta for the global admin notice is created.
+	 */
+	function test_low_max_age_admin_notice_user_meta() {
```
should have test cases for both the error and warning behaviors.

pwtyler reviewed (changes requested):
one logic question, one test coverage request.

jazzsequence committed:
add one more thing to test in behat

jazzsequence committed:
add a test for the error, specifically

jazzsequence committed:
add a test for the warning

jazzsequence committed:
simplify the conditionals

jazzsequence committed:
lint

pwtyler commented on inc/admin-interface.php:
```diff
@@ -111,39 +111,59 @@ function admin_notice_old_mu_plugin() {
 	$max_age_rank = max_age_compare();
 	$current_max_age = get_current_max_age();
+
 	if (
-		$max_age_rank > 0 &&
-		$current_max_age < WEEK_IN_SECONDS &&
-		$current_max_age !== 600
+		$max_age_rank === 0 ||
+		$current_max_age >= WEEK_IN_SECONDS ||
+		$current_max_age === 600
 	) {
-		// If the current max-age value has a rank of 3 or more (10 is the highest), we'll note that it's very low.
+		return;
+	}
```
❤️ love me some guard clauses :)

jazzsequence replied:
yeah, actually simplified this more because I realized that `$max_age_rank === 0` and `$current_max_age >= WEEK_IN_SECONDS` was testing the same thing. the way the max-age rank system works is if it's greater than or equal to the default/recommendation, then it will always be 0 and the # is based on the divergence from that recommendation.

jazzsequence committed:
specify the pages that we're on?

jazzsequence committed:
remove specific element

jazzsequence committed:
fix typo

jazzsequence committed:
take out the dashboard main page goto

jazzsequence committed:
validate the setting was saved

jazzsequence committed:
let's just remove the 5 days notice

jazzsequence committed:
just remove the test
we have similar things being tested in phpunit

jazzsequence committed:
we can actually remove current_max_age >= WEEK_IN_SECONDS
this is covered by max_age_rank. if max_age_rank is 0 it, by definition, means that the current max age is >= WEEK_IN_SECONDS

jazzsequence committed:
if we're only comparing two things, we don't need to break the if clauses up

jazzsequence committed:
we do want to warn users if they have manually set their max age to 600
we just don't want to warn them if we're about to (or just did) change it to a week for them
